### DISCUSSION TITLE
How to configure the screen lock and sleep idle timeout?

### DISCUSSION THREAD

**Discussion by jnslmk** [Manjaro-Sway/manjaro-sway#844] [Q&A]
Can someone point me in the right direction on how to configure the timeouts for screen locking and sleep?

Thanks a lot!

**Comment by trinitronx** [comment]
Recently it looks like there were [changes to how `swayidle` is launched and where its configuration lives][1].  Previously, it used to be located in `/etc/sway/autostart` and was using some sway config variables (`$swayidle`, `$idle_timeout`, `$locking_timeout`, `$screen_timeout`, `$sleep_timeout_bat`, `$sleep_timeout_ac`, `$sleep_delay`).  In that older version of the `manjaro-sway-settings` package, you could simply override these variables in your home directory's Sway configs (e.g. `~/.config/sway/config`).

Now, some things have changed in a recent update.

Launching `swayidle` has been moved into a [SystemD user unit in `/usr/lib/systemd/user/swayidle.service`][2].  That SystemD unit (among others) was moved into a package named `sway-services` in the `manjaro-sway` `pacman` repo.  It now launches a python wrapper script named [`/usr/bin/swayidle-conf`][3] which manages launching the actual `swayidle` binary using arguments that are now stored in a YAML file: [`/etc/sway/idle.yaml`][4].

The new `swayidle-conf` script also checks in the home directory location for a YAML file `~/.config/sway/idle.yaml`, which gets merged with PyYAML's `.update()` method.  So, now we can override settings by placing that file in our home directory and overriding the top-level YAML keys.  Note: the PyYAML `.update()` method does not do a "deep merge", so only the top-level keys are merged, lists are overridden completely rather than appended, etc...  That's a good thing, because it allows us to override the defaults.

Finally, I've noticed that ~~`swayidle` supports multiple "events" keys (e.g. 2x `before-sleep`/`after-resume` keys, etc...)~~ [^1], but the current implementation in `swayidle-conf` currently does not support that.  ~~So, I've made [this pull request][5] to add that feature.~~ [^2]  ~~If that one merges, we can use a top-level `events` key to list our event overrides which could contain multiple of the same event key (like [this example in my dotfiles][6]).~~

[^1]: It turns out [this was not actually the case][7], but simply just appeared to work while silently overriding and using the last event passed.
[^2]: This PR has now been closed

[1]: https://github.com/manjaro-sway/desktop-settings/commit/8186fab31e43980b4e405cc8153d1ad73245c0cb
[2]: https://github.com/xdbob/sway-services/blob/e3d9b8b390dc12224bdca23a0eb78265c824a360/systemd/swayidle.service.in
[3]: https://github.com/xdbob/sway-services/blob/e3d9b8b390dc12224bdca23a0eb78265c824a360/bin/swayidle-conf
[4]: https://github.com/manjaro-sway/desktop-settings/blob/2661735984feb25ca00fd911199ab75bcf6a15f9/community/sway/etc/sway/idle.yaml
[5]: https://github.com/xdbob/sway-services/pull/32
[6]: https://github.com/trinitronx/dotfiles/blob/9bed7c68b3f1b037043c86c62773f512da9e60a3/dot_config/sway/idle.yaml#L33-L37
[7]: https://github.com/xdbob/sway-services/pull/32#issuecomment-2638347633

**Reply by boredland** [reply]
@trinitronx we could also apply your PR as a patch, if that is urgent.

**Reply by trinitronx** [reply]
Thanks!  It's not super urgent for me at least.  I just built a local `PKGBUILD` and installed that in the meantime.  Not sure if @xdbob is watching the PRs closely, but we can either wait or if you think it's worth doing a patch then go for it 🤷

**Reply by trinitronx** [reply]
@boredland:
> @trinitronx we could also apply your PR as a patch, if that is urgent.

I've just closed that PR based on testing and [new information][1].  (**_TL;DR:_** `swayidle` doesn't actually support multiple events, but just appeared to based on logs)

FWIW: I've also managed to get a proof-of-concept working for `lock` / `unlock` integration with `gnome-keyring-daemon`, and setting SystemD/`logind`'s `LockedHint` via D-Bus.  The `lock` and `unlock` events can be triggered by `loginctl lock-session` / `loginctl unlock-session`.  I set `$locking` to just trigger via `loginctl lock-session` so it works via <kbd>$mod</kbd>+<kbd>Shift</kbd>+<kbd>e</kbd>, <kbd>l</kbd> and `swayidle` timeout triggers.  [Example working configuration here][2]. (Depends on custom `/etc/pam.d/swaylock` config & `pam_exec.so`)

[1]: https://github.com/xdbob/sway-services/pull/32#issuecomment-2638347633
[2]: https://github.com/trinitronx/dotfiles/pull/98
